### PR TITLE
feat: add setters/getters, handling default params

### PR DESCRIPTION
**Pull request by pcolaianni**
connects #6
An attempt to solve this: https://github.com/biosustain/OpenMS/issues/6
I need sane default values for all those parameters/members.
I'm still not sure about the best way to copy values (or parameter entirely) from members to `param_`.

I've added also some spectrum data into the test file. For now it's not used.

The tests all pass, except the one about `getParameters()` because copying from members into `param_` is not implemented, yet.

### CODE REVIEW

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h [hunk 1]
```diff
@@ -47,18 +49,44 @@ namespace OpenMS
...
-  class OPENMS_DLLAPI SpectrumExtractor
+  class OPENMS_DLLAPI SpectrumExtractor :
+    public DefaultParamHandler
   {
 public:
     SpectrumExtractor();
-    ~SpectrumExtractor();
+    virtual ~SpectrumExtractor();
+
+    void setRTWindow(double rt_window);
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: small change to allow for pass by reference:  double& rt_window

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h [hunk 2]
```diff
@@ -47,18 +49,44 @@ namespace OpenMS
...
+    double getRTWindow() const;
+
+    void setMinScore(double min_score);
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: change to pass by reference.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h [hunk 3]
```diff
@@ -47,18 +49,44 @@ namespace OpenMS
...
+    double getMinScore() const;
+
+    void setMinForwardMatch(double min_forward_match);
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: change to pass by reference.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h [hunk 4]
```diff
@@ -47,18 +49,44 @@ namespace OpenMS
...
+    double getMinForwardMatch() const;
+
+    void setMinReverseMatch(double min_reverse_match);
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: change to pass by reference.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h [hunk 5]
```diff
@@ -47,18 +49,44 @@ namespace OpenMS
...
+    double getMinReverseMatch() const;
+
+    void setMZTolerance(double mz_tolerance);
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: change to pass by reference.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h [hunk 6]
```diff
@@ -47,18 +49,44 @@ namespace OpenMS
...
+    double getMZTolerance() const;
+
+    void setMZToleranceUnits(String mz_tolerance_units);
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: change to pass by reference.

File: src/openms/source/ANALYSIS/OPENSWATH/SpectrumExtractor.cpp [hunk 7]
```diff
@@ -36,10 +36,97 @@
...
-    // initialize variables?
+    getDefaultParameters(defaults_);
+
+    // write defaults into Param object param_
+    defaultsToParam_();
+  }
+
+  SpectrumExtractor::~SpectrumExtractor() {}
+
+  void SpectrumExtractor::setRTWindow(double rt_window)
```
Comment (dmccloskey) [inline src/openms/source/ANALYSIS/OPENSWATH/SpectrumExtractor.cpp]: Update to match .h pass by reference.  And same for other functions below.

File: src/openms/source/ANALYSIS/OPENSWATH/SpectrumExtractor.cpp [hunk 8]
```diff
@@ -36,10 +36,97 @@
...
+    min_forward_match_ = (double)param_.getValue("min_forward_match");
+    min_reverse_match_ = (double)param_.getValue("min_reverse_match");
+    mz_tolerance_ = (double)param_.getValue("mz_tolerance");
+    mz_tolerance_units_ = (String)param_.getValue("mz_tolerance_units");
+  }
+
+  void SpectrumExtractor::getDefaultParameters(Param& params)
+  {
+    params.clear();
+    // TODO also set min and max values for these defaults (eg. params.setMinFloat(...))
```
Comment (dmccloskey) [inline src/openms/source/ANALYSIS/OPENSWATH/SpectrumExtractor.cpp]: Good idea.

File: src/tests/class_tests/openms/source/SpectrumExtractor_test.cpp [hunk 9]
```diff
@@ -62,6 +289,71 @@ START_SECTION(~SpectrumExtractor())
...
+}
+END_SECTION
+
+START_SECTION(getParameters().getDescription("rt_window"))
+{
+  TEST_EQUAL(ptr->getParameters().getDescription("rt_window"), "Retention time window.")
+}
+END_SECTION
+
+delete ptr;
```
Comment (dmccloskey) [inline src/tests/class_tests/openms/source/SpectrumExtractor_test.cpp]: Very thorough.  This looks very nice so far.

File: src/tests/class_tests/openms/source/SpectrumExtractor_test.cpp [hunk 10]
```diff
@@ -45,6 +46,232 @@ using namespace std;
 START_TEST(SpectrumExtractor, "$Id$")
 
 /////////////////////////////////////////////////////////////
+// Dummy peak data
+
+MSSpectrum spectrum;
+spectrum.resize(73);
+MSSpectrum::Iterator it = spectrum.begin();
+
+it->setMZ(61.92);
```
Comment (dmccloskey) [inline src/tests/class_tests/openms/source/SpectrumExtractor_test.cpp:58]: Add a note on where this data is coming from in a multiline comment
// Raw spectrum data acquired in DDA mode (i.e., product ion full spectrum scan)
// measured on a QTRAP 5500 corresponding to C-Aconitate
// taken from E. coli grown on glucose M9 during steady-state 
// for flux analysis.

**Review by dmccloskey — COMMENTED** [review]
Looks good so far!   A few minor changes.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h [hunk 11]
```diff
@@ -47,18 +49,44 @@ namespace OpenMS
...
+    String getMZToleranceUnits() const;
+
+    void getDefaultParameters(Param& params);
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: You can add in a function e.g., 
void getParameters (Param& params)
{
getDefaultParameters(params);
params.setValue("rt_window",rt_window_);
...
}
that updates the params object before returning it to the user.  

Alternatively, you could add a member "Param params_;" that stores the params_ object that is updated in each setter method.
In that case, getParameters would look like the following: 

void getParameters(Param& params)
{
params = params_
}

Comment (pcolaianni) [reply src/openms/include/OpenMS/ANALYSIS/OPENSWATH/SpectrumExtractor.h]: opened an issue: https://github.com/biosustain/OpenMS/issues/16

**Pull request closed without merging**